cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+	// perOperationBalance maintains per-operation balances of withdrawn credits
+	perOperationBalance map[string]float64
+	// updateTime is the last time client was updated
+	updateTime time.Time
+}
+
+func newClient(id string, currentTime time.Time) *client {
+	return &client{
+		id:                  id,
+		updateTime:          currentTime,
+		perOperationBalance: map[string]float64{},
+	}
+}
+
+// Spend depletes the balance from the client for the given operation. The usage
+// would be when a client submits spans for operation X, it is spending those
+// credits it requested for operation X.
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance := c.perOperationBalance[operationName]
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	c.perOperationBalance[operationName] = balance
+	return true
+}
+
+// CreditAccruerOptions provides values to be used with a creditAccruer object
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
+}
+
+type creditAccruer struct {
+	options                 CreditAccruerOptions
+	serviceName             string
+	perOperationRateLimiter OverrideMap
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func newCreditAccruer(serviceName string, options CreditAccruerOptions, timeNow func() time.Time) *creditAccruer {
+	defaultRateLimiter := newTokenBucket(options, timeNow)
+	perOperationRateLimiter := NewOverrideMap(options.MaxOperations, defaultRateLimiter)
+	return &creditAccruer{
+		options:                 options,
+		serviceName:             serviceName,
+		updateTime:              timeNow(),
+		timeNow:                 timeNow,
+		perOperationRateLimiter: perOperationRateLimiter,
+	}
+}
+
+// Withdraw credits from the creditAccruer for the given operationName with an
+// upper limit of maxWithdrawal.
+func (accruer *creditAccruer) Withdraw(operationName string, maxWithdrawal float64) float64 {
+	var rateLimiter *tokenBucket
+	if accruer.perOperationRateLimiter.Has(operationName) ||
+		accruer.perOperationRateLimiter.IsFull() {
+		rateLimiter = accruer.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(accruer.options, accruer.timeNow)
+		accruer.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// CreditStore manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type CreditStore interface {
+	io.Closer
+	// Withdraw deducts as many credits from the service on behalf of the client
+	// without exceeding MaxClientBalance. The client is identified by a unique
```

Comment:
nit: MaxClientBalance is not part of the interface, so unclear what it means

Reply:
Right, going to move to struct anyway. Maybe can clarify better there.